crates/pathfinder/src/rpc/api.rs

```diff
@@ -331,6 +333,53 @@ impl RpcApi {
             .and_then(|x| x)
     }
 
+    /// Fetches [GlobalRoot] of a block by block hash.
+    ///
+    /// Tries storage first, then the sequencer.
+    async fn get_root_by_hash(&self, block_hash: StarknetBlockHash) -> RpcResult<GlobalRoot> {
+        let storage = self.storage.clone();
+
+        let handle = tokio::task::spawn_blocking(move || {
+            let mut connection = storage
+                .connection()
+                .context("Opening database connection")
+                .map_err(internal_server_error)?;
+
+            let transaction = connection
+                .transaction()
+                .context("Creating database transaction")
+                .map_err(internal_server_error)?;
+
+            let root = StarknetBlocksTable::get_root(
+                &transaction,
+                StarknetBlocksBlockId::Hash(block_hash),
+            )
+            .context("Read block from database")
+            .map_err(internal_server_error)?;
+
+            Ok(root)
+        });
+
+        let root = handle
+            .await
+            .context("Database read panic or shutting down")
+            .map_err(internal_server_error)
+            // flatten is unstable
+            .and_then(|x| x)?;
+
+        match root {
+            Some(root) => Ok(root),
+            None => Ok(self
+                .sequencer
+                .block_by_hash(BlockHashOrTag::Hash(block_hash))
+                .await
+                .context("Fetch block from sequencer")
+                .map_err(internal_server_error)?
+                .state_root
+                .expect("State root is always present for non pending block")),
+        }
```

Comment:
I would move this into the calling function. Automatically querying the sequencer is a weird thing in general -- so having it be part of an innocent looking function is dangerous.

Reply:
I am also a bit lost on the **why** is this done?

Reply:
> I am also a bit lost on the why is this done?
The sequencer reply does not provide this field and oftentimes we won't have it in the db anyway because sync is not at head.

Reply:
Oki after understanding why, I can't really see any way out of this. It's not great but if we have to provide this api and there's no other way. I would adjust the `.context` for this nested thing so that at least we will know.

Reply:
The only way out of this would be Starkware providing this information on their API (for pending blocks). It's not great to have to do this, but I don't have a better idea...

Reply:
Yeah, I totally understand Mirko's - `pending` s a temporary solution so let's just forward what we get from the sequencer, but at the same time we are bound by the spec, and the structure there does not match 1:1, we do some transformations.